core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -2946,14 +3040,170 @@ private void scheduleAsyncPersistenceInternal(LockedInodePath inodePath) throws
+              ufsChildFingerprints.put(ufsChildStatus.getName(),
+                  Fingerprint.create(Fingerprint.getUfsName(ufs), ufsChildStatus)
+                      .serialize());
+            }
+            InodeDirectory inodeDir = (InodeDirectory) inode;
+            for (Inode<?> child : inodeDir.getChildren()) {
+              inodeChildren.put(child.getName(), child);
+            }
+
+            // Iterate over ufs children and process children which do not exist in Alluxio.
+            for (Map.Entry<String, String> ufsEntry : ufsChildFingerprints.entrySet()) {
+              if (!inodeChildren.containsKey(ufsEntry.getKey()) && !PathUtils
+                  .isTemporaryFileName(ufsEntry.getKey())) {
+                // Ufs child exists, but Alluxio child does not. Must load metadata.
+                loadMetadata = true;
+                break;
+              }
+            }
+
+            // Iterate over Alluxio children and process persisted children.
+            for (Map.Entry<String, Inode<?>> inodeEntry : inodeChildren.entrySet()) {
+              if (!inodeEntry.getValue().isPersisted()) {
+                // Ignore non-persisted inodes.
+                continue;
+              }
+
+              String ufsFingerprint = ufsChildFingerprints.get(inodeEntry.getKey());
+              boolean deleteChild =
+                  !UfsSyncUtils.inodeUfsIsSynced(inodeEntry.getValue(), ufsFingerprint);
+
+              if (deleteChild) {
+                TempInodePathForDescendant tempInodePath =
+                    new TempInodePathForDescendant(inodePath);
+                tempInodePath.setDescendant(inodeEntry.getValue(),
+                    inodePath.getUri().join(inodeEntry.getKey()));
+
+                try {
+                  deleteInternal(tempInodePath, false, System.currentTimeMillis(),
+                      syncDeleteOptions, journalContext);
+                } catch (DirectoryNotEmptyException | IOException e) {
+                  // Should not happen, since it is an unchecked delete.
+                  LOG.error("Unexpected error for unchecked delete. error: {}", e.toString());
+                }
+                // Must load metadata afterwards.
+                loadMetadata = true;
+              }
+            }
+          }
+        }
+      }
+    } finally {
+      if (deletedInode) {
+        // If the inode was deleted, then the inode path should reflect the delete.
+        inodePath.unlockLast();
+      } else {
+        // If the inode was not deleted, downgrade the last inode lock, if necessary.
+        inodePath.downgradeLastWithScheme(lockingScheme);
+      }
+    }
+
+    if (loadMetadata) {
+      try {
+        loadMetadataAndJournal(inodePath,
+            LoadMetadataOptions.defaults().setCreateAncestors(true).setLoadDirectChildren(true),
+            journalContext);
+      } catch (Exception e) {
+        LOG.debug("Failed to load metadata for path: {}", inodePath.getUri());
```

Comment:
Is this an expected case and should we still notify the cache if this happens?

Reply:
Thanks. It no longer updates the cache now.